R-packages/modeltools/R/response.R

```diff
@@ -0,0 +1,80 @@
+library(assertthat)
+
+#' Compute the corresponding observed response values for a set of signals.
+#'
+#' The response for an observed data point at time `t` is defined as the mean of observed values on
+#' days `t + s` and `t + e` multiplied by `e - s`, where `s` and `e` are the number of days between
+#' `forecast_date` and the start and end, respectively, of the next full `incidence_period`.  We
+#' choose to take the mean and multiply by the incidence period length rather than simply sum the
+#' observed values in order to avoid treating missing values as 0.
+#'
+#' @param signals Wide data frame of signals.  We expect the following columns:
+#'     \itemize{
+#'     \item{`geo_value`}{Strings of geographic locations.}
+#'     \item{`time_value`}{Dates of training data.}
+#'     \item{Covariate columns}{Columns with names of the form `value-{days}:{signal}` whose values
+#'         correspond to `{signal}` `{days}` before `time_value`}
+#'     }
+#' @param response_name Name of response signal.  The value of the response variable on a single
+#'     day should be located in column `value-0:{response_name}` of `signals`.
```

Comment:
No I am enforcing the nomenclature that `value-{n}` corresponds to covariate values while `value+{n}` corresponds to the response.  This is pulled from `quantgen_forecaster()` [here](https://github.com/cmu-delphi/covidcast/blob/edf101b4beb23b9a29acda0e1c76a49b8c679bd5/R-packages/modeltools/R/quantgen.R#L146).  I think it is a good convention since it:
1.  makes it easy in code to pick out all the covariates but not the responses (or vice versa).
2.  allows `value+0` to have a meaning for, e.g., nowcasting that is distinct from the observed values.

Reply:
I agree it's a good convention. In `quantgen_forecaster()`, it makes the lagged df using `covidcast::aggregate_signals()`. But I don't think `aggregate_signals()` can give you `value-0`. For example:

```
signals <- tibble::tibble(
  data_source = "usa-facts",
  signal = c("confirmed_incidence_num", "deaths_incidence_num"))
raw_df <- covidcast::covidcast_signals(data_source = signals$data_source,
                                       signal = signals$signal,
                                       start_day = "2021-02-01",
                                       end_day = "2021-02-07",
                                       geo_type = "state")
names(covidcast::aggregate_signals(raw_df, dt = 0))
names(covidcast::aggregate_signals(raw_df, dt = -0))
names(covidcast::aggregate_signals(raw_df, dt = +0))
```
On my laptop, all three `names()` calls give me `value+0`.

Reply:
Ah, I see, I misinterpreted that were actually adding a "-" to the front rather than multiplying by -1.

I can simply change this to expect `value+0:{name}` to be the covariate, or I can go one step further and change the naming convention for the responses from `value+{ahead}:{name}` to something like `response+{ahead}:{name}`, or do something else you think would be better than both of those.

Reply:
Also this mistake was made in #494 so I will need to change it there once we decide on something.

Reply:
I personally like the convention you first assumed. I just don't know the best way to create/allow for `value-0:{name}`, as I'm not sure if other functions in `evalcast`/`covidcast` make assumptions about the column names. Perhaps @dajmcdon or @huisaddison can comment on this.

Reply:
Nothing in `evalcast` depends on this at all.  We should also assume that `covidcast` is immutable.  If we really want to make this work we could have a wrapper around `aggregate_signals()` that changes the name although that is quite inelegant.  It might become less bad if we also implement in this wrapper the ability to lag long or wide data frames or columns derived from the raw downloaded signals (all not possible in `covidcast_signals()`).

Reply:
Hey all -- if I understand correctly, the proposed options are:

1. Enforce the `-0` is a covariate, `+0` is a response convention by extra code that is "inelegant"
2. Have separate `value-{...}` and `response+{...}` column names.

I personally prefer the latter.  I think that the former is more elegant in the abstract, but I don't think it's worth it because (a) the code, if it's clunkier, would be harder to understand; (b) the -0 vs +0 distinction is defined purely by us, and if anybody writes code that munges the column names and converts the "numeric substring" into a numeric, the +0 vs -0 distinction will be lost; and (c) using `response` in the column name explicitly is more verbose, and hence easier to understand for new users, than the convention that positive offsets correspond to a response.

Reply:
@huisaddison this is good reasoning.  Implemented the latter.

Reply:
Thanks @sgsmob, this looks good to me.